### ISSUE TITLE
Small holes in the created polygon with 2 adjacent census tracts, causes PCR_1 fail

### ISSUE THREAD

**Issue by chandanahs** [Wireless-Innovation-Forum/Spectrum-Access-System#814]
when the service area is created in PCR_1 using 2 PALs with adjacent census tracts (pal_record_a['fipsCode'] = 20063955100 & pal_record_b['fipsCode'] = 20063955200), created Polygon after cascading is not the outer polygon, but has very small gap between census tracts.

When adjacent sides of adjacent census tracts are not having the exact same co-ordinates, we see very small gap which is considered as holes and PPA created during PCR_1 is failing with "AssertionError: PPA Zone is not within service area" as there is tiny hole in cascaded polygon using 2 given PALs with adjacent census tracts.

the similar scenario is discussed here: [https://gis.stackexchange.com/questions/277334/shapely-polygon-union-results-in-strange-artifacts-of-tiny-non-overlapping-area](url)

So Pls check if we have to change the test code as well as reference model code to create a outer polygon without any tiny holes!

Have a look at the attachments.

[raw_map.zip](https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/files/2435248/raw_map.zip)

![adjacentcensustractscascadedpolygonwithtinyholes](https://user-images.githubusercontent.com/42067597/46309897-c15ca380-c58b-11e8-8725-ef724d9a742c.PNG)

**Comment by chandanahs** [comment]
@kate-harrison : Could you have a look? or redirect to the right person?

**Comment by chandanahs** [comment]
i have observed holes with less than 6 meter width holes between 2 adjacent census tracts.

attaching the screen shot for the same. Yes, Buffering would solve this problem.

![tinyholebtw2censustracts](https://user-images.githubusercontent.com/42067597/46692904-66076280-cbd6-11e8-83ae-89523aa433fd.PNG)

def isPpaWithinServiceArea(pal_records, ppa_zone_geometry):
   census_tracts_for_pal = [
      utils.ToShapely(drive.census_tract_driver.GetCensusTract(
          pal['license']['licenseAreaIdentifier'])
                      ['features'][0]['geometry']) for pal in pal_records]
  pal_service_area = ops.cascaded_union(census_tracts_for_pal)

  **pal_service_area = geometry.Polygon(pal_service_area.exterior)
  pal_service_area = pal_service_area.buffer(1e-6,1,join_style=JOIN_STYLE.mitre).buffer(-1e-6,1,join_style=JOIN_STYLE.mitre)**
  # Convert GeoJSON dictionary to Shapely object.
  ppa_zone_shapely_geometry = utils.ToShapely(ppa_zone_geometry)

  return ppa_zone_shapely_geometry.buffer(-1e-6).within(pal_service_area)

above solved the problem.

**Comment by chandanahs** [comment]
I have in the process of getting Contribution permission, I will create a Pull request once i have the permission.

**Comment by kate-harrison** [comment]
Sounds great, thank you!

**Comment by chandanahs** [comment]
i did not receive any response from Lee for write permission!

**Comment by sbdt-google** [comment]
I checked the proper json from SAS-Data (and not the one that is osbolete - see my email and #856),
 and I dont see the mismatch that you reported.

![screenshot from 2018-11-15 14-35-36](https://user-images.githubusercontent.com/30304613/48586204-2d009300-e8e4-11e8-92aa-721774ba9079.png)
![screenshot from 2018-11-15 14-36-35](https://user-images.githubusercontent.com/30304613/48586211-2f62ed00-e8e4-11e8-9e84-99578e48a506.png)

**Comment by diptivete** [comment]
@kate-harrison @sbdt-google Yes, this is not a problem upon using correct json file from SAS-Data repo. 
Guess, we used older 20063955200.json file under /data/census_tracts/ directory that resulted in this confusion, sorry about that.
Nevertheless, I tried all 7 PCR cases with and without changes made in PR #852 on @chandanahs behalf as she is currently OoO.
They are working for WINNF default configs.
